src/state/delete-text.action.ts

```diff
@@ -0,0 +1,185 @@
+import { Action } from './action';
+import { ISegment } from '../entities/segment';
+import { ICursor } from '../entities/cursor';
+import { getCurrSegment, IEditorState, findSegment } from './editor.state';
+import { remove } from '../common/array';
+
+export class DeleteTextAction implements Action {
+
+    private rematchIndex(segments: ISegment[]): ISegment[]{
+        return segments.map((currSegment: ISegment, idx: number) => {
+            return Object.assign<any, ISegment, Partial<ISegment>>(
+                {},
+                currSegment,
+                {
+                    index: idx
+                }
+            );
+        });
+    }
+
+    private deleteSingleChar(state: IEditorState): IEditorState {
+        const { cursor, segments } = state;
+        let newState: IEditorState = state;
+        let newSegments: ISegment[] = segments;
+        let newCursor: ICursor = cursor;
+
+        const segment = getCurrSegment(newState);
+        if (!segment) {
+            return state;
+        }
+
+        const newSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            segment,
+            {
+                // to remove the single char at the cursor
+                content: remove<string>(
+                    segment.content,
+                    newCursor.startOffset - 1
```

Comment:
```suggestion
                    segment.content,
                    newCursor.startOffset - 1,
                    1
```
Prefer making thing explicit to avoid confusion.